jormungandr/src/blockchain/storage.rs

```diff
@@ -37,205 +113,306 @@ struct BlockIterState {
     pending_infos: Vec<BlockInfo<HeaderHash>>,
 }
 
-impl Storage {
+impl Storage03 {
     pub fn new(storage: NodeStorage) -> Self {
-        Storage {
-            read_connection: storage.clone(),
-            write_connection_lock: Lock::new(storage),
-        }
+        let mut rt = runtime::Builder::new()
+            .name_prefix("new-storage-worker-")
+            .core_threads(1)
+            .build()
+            .unwrap();
+
+        rt.block_on_std(async move {
+            let manager = ConnectionManager::new(storage);
+            let pool = Pool::builder().build(manager).await.unwrap();
+            let write_lock = Arc::new(Semaphore::new(1));
+
+            Storage03 { pool, write_lock }
+        })
     }
 
-    #[deprecated(since = "new blockchain API", note = "use the stream iterator instead")]
-    pub fn get_inner(&self) -> impl Future<Item = NodeStorage, Error = StorageError> {
-        future::ok(self.read_connection.clone())
+    async fn run<F, R>(&self, f: F) -> Result<R, StorageError>
+    where
+        F: FnOnce(&mut NodeStorageConnection) -> Result<R, StorageError> + Send + 'static,
+        R: Send + 'static,
+    {
+        run_blocking_with_connection(&self.pool, f).await
     }
 
-    pub fn get_tag(
-        &self,
-        tag: String,
-    ) -> impl Future<Item = Option<HeaderHash>, Error = StorageError> {
-        future::result(self.read_connection.get_tag(&tag))
+    pub async fn get_tag(&self, tag: String) -> Result<Option<HeaderHash>, StorageError> {
+        self.run(move |connection| connection.get_tag(&tag)).await
     }
 
-    pub fn put_tag(
-        &mut self,
-        tag: String,
-        header_hash: HeaderHash,
-    ) -> impl Future<Item = (), Error = StorageError> {
-        let mut write_connection_lock = self.write_connection_lock.clone();
-
-        future::poll_fn(move || Ok(write_connection_lock.poll_lock())).and_then(move |mut guard| {
-            match guard.put_tag(&tag, &header_hash) {
-                Err(error) => future::err(error),
-                Ok(res) => future::ok(res),
-            }
-        })
+    pub async fn put_tag(&self, tag: String, header_hash: HeaderHash) -> Result<(), StorageError> {
+        let _ = self.write_lock.acquire().await;
```

Comment:
Is it still necessary to lock sqlite for mutative operations? Is it some sort of a way to avoid transaction conflicts?

Reply:
We had cases when write operations were done in the wrong order. In particular, a block could've been written before its parent. Introducing this lock helped to resolve the issue. I think the issue appeared due to wrong scheduling of futures involving write operations: operations are not necessarily done in the same order they were requested.

Reply:
If consistency is not verified on the database level, I think it's still possible for the client threads to put blocks in the wrong order. The lock just makes it less likely at a cost of extra synchronization.

Reply:
This will do for now; I have added https://github.com/input-output-hk/chain-libs/issues/276 to resolve the consistency problems in chain-storage-sqlite.